Add iSCSI targets/LUNs through Pacemaker clusters + development guidelines and tooling

mpeterson opened the pull request:
* Support pcs for zfs-generic to configure iSCSI targets and LUNs. It uses the targetcli implementation since it's the default pcs resources and also to keep compatibility with the rest of the code.
* Adds devcontainer and vagrant, as well as `CONTRIBUTING.md` guidelines to make it easier to develop
* Simplifies the `ci/bin/run.sh` to not depend on `uuidgen`

closes #462

mpeterson committed:
Add iSCSI targets/LUNs through Pacemaker clusters

Support pcs for zfs-generic to configure iSCSI
targets and LUNs. It uses the targetcli implementation
since it's the default pcs resources and also to
keep compatibility with the rest of the code.

closes #462

mpeterson commented:
@travisghansen this is still a WIP but to check alignment. 

I'm also looking to add it to the CI gate but it seems that the server where it SSHs to is not configured by files in the repository. For the CI gate to be able to check this strategy it would need to have pcs ready to be used.

mpeterson commented on src/driver/controller-zfs-generic/index.js:
```diff
@@ -298,8 +298,70 @@ create /backstores/block/${assetName}
+              pcs_group = this.options.iscsi.shareStrategyPcs.pcs_group;
+
+              let groupText = `group ${pcs_group}`;
+              let createTargetText = [
```
yeah, that's how I wanted to do it to start with, as it's much cleaner. Then I looked at the implementation of `targetCliCommand` and how the underlying `exec` works and I see that it ends up doing just the one string. I'll separate it for future proofing though.

mpeterson commented on src/driver/controller-zfs-generic/index.js:
```diff
@@ -851,6 +962,56 @@ save_config filename=${this.options.nvmeof.shareStrategySpdkCli.configPath}
+      logCommand += " ";
+      
+      if (term.startsWith("incoming_password=")) {
+        logCommand += "incoming_password=<redacted>";
```
not supported by the `pcs` resource agent

mpeterson commented on src/driver/controller-zfs-generic/index.js:
```diff
@@ -851,6 +962,56 @@ save_config filename=${this.options.nvmeof.shareStrategySpdkCli.configPath}
+      pty: true,
+    };
+    let response = await execClient.exec(
+      execClient.buildCommand(command, args),
```
I agree with you, but ended going how the other `*Command` were handling it. And now reviewing the code again part of the reason for this to be done that way is because it's actually because they are doing `sh -c '<command_string>'`, which has the advantage of forcing a specific shell and not being environment dependent.

mpeterson commented:
I addressed the comments you put and left the conversations unresolved to let you close the items if you agree. 

Regarding your questions:

>  Has it been tested? 

Not yet, I didn't know how to build it properly for testing. I saw your comment on the issue and will try to follow it, but seems complicated. I'm thinking this project could benefit from a [devcontainer](https://devcontainers.github.io/) and a way to run tests or build the image and deploy it in a cluster for testing.

Is this just me? How have other contributors tested their code? 

> Is the idea that the driver is configured to exec commands over ssh to the vip such that if one of the nodes dies it will land on the other?

Correct, the SSH target will be the VIP and then `pcs` will handle replication to the other nodes or reconciliation when one or more is unavailable during configuration (as long as it is possible by current quorum).

mpeterson committed:
Small CR and preparatory work for CI

mpeterson commented:
> > @travisghansen this is still a WIP but to check alignment.
> > I'm also looking to add it to the CI gate but it seems that the server where it SSHs to is not configured by files in the repository. For the CI gate to be able to check this strategy it would need to have pcs ready to be used.
> 
> Yeah, the CI runs in my homelab due to the need for various servers/etc. If we want to add this to CI it may be good to have a call/screenshare and go through a setup process to get everything going.

it's quite simple, you'll need at least 2 nodes for this. I'll give you a rough idea here of the [actual setup](https://github.com/ewwhite/zfs-ha/issues/40#issuecomment-2661040698) I have, and if you want we can hop on a call. These commands are for RHEL-flavored Linuxes. In particular I did it with AlmaLinux.

```sh
# IMPORTANT: All to be executed as root

firewall-cmd --permanent --add-service=high-availability
firewall-cmd --reload

# I'm assuming you have a working ZFS pool, so I'm not going to put all commands related to that. Let me know if you need them

dnf config-manager --set-enabled highavailability
dnf install -y pcs fence-agents-all
systemctl enable pcsd
systemctl enable corosync
systemctl enable pacemaker
systemctl start pcsd

cd /usr/lib/ocf/resource.d/heartbeat/
wget https://github.com/skiselkov/stmf-ha/raw/master/heartbeat/ZFS
chmod +x ZFS

pcs host auth nas-01 nas-02
pcs cluster setup zfs-cluster --start --enable nas-01 addr=10.3.25.11 nas-02 addr=10.3.25.12

pcs property set no-quorum-policy=ignore

# adjust the path of importargs to where your disks are, in my case I'm using mapper to have multipath availability
pcs resource create --future nas-zfs 'ocf:heartbeat:ZFS' pool="nibbler" importargs="-d /dev/mapper/" op start timeout="90" op stop timeout="90" group group-nas

# this is the VIP
pcs resource create --future nas-ip 'ocf:heartbeat:IPaddr2' ip=10.3.0.10 cidr_netmask=24 group group-nas

pcs resource defaults update resource-stickiness=100

# make the targetcli available for the iSCSI resource agent 
dnf install targetcli

## if you want to test manually to have the iSCSI configured

# create a volume to use, example

zfs create -V 100M tank/.k8s/tester

pcs resource create --future target-tester 'ocf:heartbeat:iSCSITarget' implementation="lio-t" iqn="iqn.2025-03.com.example:tester" incoming_username='foo' incoming_password='bar' group group-nas

# adjust path in your case
pcs resource create --future lun-tester 'ocf:heartbeat:iSCSILogicalUnit' implementation="lio-t" target_iqn="iqn.2025-03.com.example:tester" lun="1" path="/dev/zvol/tank/.k8s/tester" group group-nas

zfs destroy tank/.k8s/tester

pcs resource delete lun-tester
pcs resource delete target-tester
```

mpeterson committed:
Adding some sleeps to allow things to settle down

mpeterson committed:
small fix

mpeterson commented:
okey, got it how to test it. I'll test it next time I can work on this a bit.

d-uzlov commented:
> I'm thinking this project could benefit from a [devcontainer](https://devcontainers.github.io/) and a way to run tests or build the image and deploy it in a cluster for testing.

> How have other contributors tested their code?

I just build the docker image, replace the image in the helm values or deployment config, and just run my build in the cluster.

```bash
docker build . --push -t docker.io/$docker_username/$docker_repo:$docker_tag
```

Example helm values modifications:

```yaml
controller:
  driver:
    image: docker.io/daniluzlov/k8s-snippets:dcsi-proxy2-v47
node:
  driver:
    image: docker.io/daniluzlov/k8s-snippets:dcsi-proxy2-v47
```

mpeterson committed:
Added vagrant and devcontainers + fixed all tests

mpeterson commented:
> > I'm thinking this project could benefit from a [devcontainer](https://devcontainers.github.io/) and a way to run tests or build the image and deploy it in a cluster for testing.
> 
> > How have other contributors tested their code?
> 
> I build the docker image, replace the image in the helm values or deployment config, and just run my build in the cluster.
> 
> ```shell
> # run in the root of the repository
> docker build . --push -t docker.io/$docker_username/$docker_repo:$docker_tag
> ```
> 
> Example helm values modifications:
> 
> ```yaml
> controller:
>   driver:
>     image: docker.io/daniluzlov/k8s-snippets:dcsi-proxy2-v47
> node:
>   driver:
>     image: docker.io/daniluzlov/k8s-snippets:dcsi-proxy2-v47
> ```

that's awesome, we can add that also to the CONTRIBUTING.md

mpeterson committed:
Minor changes

mpeterson commented:
@travisghansen oh, since I modified the CI to also test this I added the new CI test commented out with a `TODO` until it's actually setup on your on-prem servers.

mpeterson committed:
CR fixes

mpeterson commented:
> Thanks for all the effort. I have merged a couple things to `next` so make sure to pull those in and resolve any conflicts etc.

merged with `next` and addressed comments

mpeterson committed:
remove unneeded rm

mpeterson committed:
fix vagrant provisioning script

mpeterson commented:
> Great. Shall we attempt a call and try to setup a pair of nodes to run pcs on for ci? Or even a single-node cluster perhaps.

sure, but don't you want perhaps to merge this first and do that over a different PR? I can open another issue and interact on that one with you.

mpeterson commented:
@travisghansen hey! wonder how you want to proceed here?

mpeterson committed:
Add build_push.sh

mpeterson committed:
small typo

mpeterson commented:
@travisghansen hey! I'd love to merge this before the upstream keeps moving forward and will require more changes. How can we proceed?

mpeterson committed:
fixed doc

mpeterson committed:
explain to put visibility public to package

mpeterson commented:
@travisghansen hey, pinging on this again

mpeterson commented:
@travisghansen I see the mail now. I was on a vacation when you replied and fell through the cracks. Let's schedule over it.

mpeterson committed:
add force to target creation to allow multiple non-unique usernames

mpeterson committed:
make build_push.sh executable

mpeterson committed:
update devcontainer image, enhance contributing guide, and implement mutex for PCS commands

mpeterson commented:
@travisghansen I just rebased and added the mutex as requested. I did some other improvements around the devcontainer, Vagrantfile and docs.

mpeterson commented:
Amazing! Took a while to merge it but glad we finally did it
